core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -744,6 +751,8 @@ class ReplicaManager(val config: KafkaConfig,
         }
 
         // map not yet verified partitions to a request object
+        // Since verification occurs on produce requests only, and each produce request has one batch per partition, we know the producer ID is transactional
```

Comment:
But do we know that the producerId is consistent among all transactional batches? Seems like we are assuming it below, but where is it verified?

Reply:
I mention this flaw in the PR description, but as discussed offline, will close the gap